### PR TITLE
Feature 1567 ai services provide information about executed tools

### LINKED ISSUE #1567: [FEATURE] AI Services: provide information about executed tools
- Add new field `List<ToolExecution> toolExecutions` to the `Result<T>` class:
```java
public class ToolExecution {
    ToolExecutionRequest toolExecutionRequest;
    Object toolExecutionResult;
}
```

_Originally posted by @langchain4j in https://github.com/langchain4j/langchain4j/issues/1248#issuecomment-2219830528_

### PR DESCRIPTION
**Pull request by hrhrng**
## Issue
Closes #1567

## Change
1. Add `ToolExecution` into the `Result` to represent the tool execution process.
2. Record ToolExecutions in `DefaultAiServices` and ultimately inject them into result.
3. Replace the openai model from GPT_3_5_TURBO_0613 to GPT_3_5_TURBO_16K in `modelsWithoutParallelToolCalling` as GPT_3_5_TURBO_0613 is deprecated. Remove Mistral model which doesn't support sequentially tool call actually.

## General checklist

- [X] There are no breaking changes
- [X] I have added unit and integration tests for my change
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [X] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

### CODE REVIEW

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolExecution.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    private ToolExecution(Builder builder) {
+        this.toolExecutionRequest = builder.toolExecutionRequest;
+        this.toolExecutionResult = builder.toolExecutionResult;
+    }
+
+    /**
+     * Returns the request of the tool execution.
+     * @return the request of the tool execution.
+     */
+    public ToolExecutionRequest toolExecutionRequest() {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolExecution.java]:
```suggestion
    public ToolExecutionRequest request() {
```

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolExecution.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        return toolExecutionRequest;
+    }
+
+    /**
+     * Returns the result of the tool execution.
+     * @return the result of the tool execution.
+     */
+    public String toolExecutionResult() {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolExecution.java]:
```suggestion
    public String result() {
```

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolExecution.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+         */
+        private Builder() {
+        }
+
+        /**
+         * Sets the {@code toolExecutionRequest}.
+         * @param toolExecutionRequest the {@code toolExecutionRequest}
+         * @return the {@code Builder}
+         */
+        public Builder toolExecutionRequest(ToolExecutionRequest toolExecutionRequest) {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolExecution.java]:
```suggestion
        public Builder request(ToolExecutionRequest request) {
```

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolExecution.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+            this.toolExecutionRequest = toolExecutionRequest;
+            return this;
+        }
+
+        /**
+         * Sets the {@code toolExecutionResult}.
+         * @param toolExecutionResult the {@code toolExecutionResult}
+         * @return the {@code Builder}
+         */
+        public Builder toolExecutionResult(String toolExecutionResult) {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolExecution.java]:
```suggestion
        public Builder result(String result) {
```

File: langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithToolsIT.java [hunk 5]
```diff
@@ -727,4 +720,119 @@ private static Map<String, Object> toMap(String arguments) {
...
+        assertThat(secondToolExecutionRequest.name()).isEqualTo("getTransactionAmount");
+        assertThat(secondToolExecutionRequest.arguments())
+                .isEqualToIgnoringWhitespace("{\"arg0\": \"T002\"}");
+
+        String firstResult = response.toolExecutions().get(0).toolExecutionResult();
+        assertThat(firstResult).isEqualTo("11.1");
+
+        String secondResult = response.toolExecutions().get(1).toolExecutionResult();
+        assertThat(secondResult).contains("22.2");
+
```
Comment (langchain4j) [inline langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithToolsIT.java:796]: please add an assertion to check that `spyChatLanguageModel` has been invoked only 1 time

File: langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithToolsIT.java [hunk 6]
```diff
@@ -727,4 +720,119 @@ private static Map<String, Object> toMap(String arguments) {
...
```
Comment (langchain4j) [inline langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithToolsIT.java:836]: please add an assertion to check that `spyChatLanguageModel` has been invoked 2 times

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolExecution.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
+package dev.langchain4j.agent.tool;
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolExecution.java]: Please move it to `dev.langchain4j.service.tool` in `langchain4j` module

**Review by ghost — COMMENTED** [review]
@hrhrng good job, thank you!

There is also a conflict, could you please check?

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolExecution.java (hunk 1 above)
Comment (hrhrng) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolExecution.java]: Perhaps modify all the fields as well?

Comment (hrhrng) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolExecution.java]: @langchain4j

Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolExecution.java]: Yes, but this is not that important as we do not expose those internal names outside

File: langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithToolsIT.java (hunk 5 above)
Comment (hrhrng) [reply langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithToolsIT.java:796]: It will be called twice here and three times in parallel.

Comment (langchain4j) [reply langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithToolsIT.java:796]: you're right!

**Review by ghost — APPROVED** [review]
@hrhrng good job, thank you!

**Pull request merged**